dataclasses_json/core.py

```diff
@@ -314,14 +314,16 @@ def _decode_generic(type_, value, infer_missing):
             type_options = _get_type_args(type_)
             res = value  # assume already decoded
             if type(value) is dict and dict not in type_options:
-                # FIXME if all types in the union are dataclasses this
```

Comment:
I don't think this is resolved by your commit, thus should be kept

Reply:
This PR explicitly targets this comment - this is what it is attempting to solve.

Instead of simply accepting the first dataclass in the list of `type_options`, each type is iterated, a decode attempt is made, and if it succeeds, we use that dataclass - otherwise we attempt to decode the next type in the options array.  In this way, we try to find the best fitting class in the `type_options` list - the best fitting match in the union.

Reply:
But still only one can fit, so this can be done in a more elegant manner with for-comp and `if` checking that the type got decoded without errors. Then you won't need this 'changed' flag and overall readability will improve. However, I leave the final decision on whether to (not) do this small improvement to you :)

Reply:
I updated this to check if the value had changed without the unneeded `changed` variable.

I also updated the warning to be more accurate.

I also added a testcase to ensure that the object remains a `dict` when decoding fails.